modules/network/dns-zone/main.bicep

```
@@ -0,0 +1,74 @@
+@description('Required. Specifies the resource name')
+param name string
+
+@description('Optional. Tags to assign to the resources used in dns-zones.')
+param tags object = {}
+
+@allowed([
+  'Public'
+  'Private'
+])
+@description('Optional. The type of this DNS zone, default set to Public.')
+param zoneType string = 'Public'
+
+@description('Optional. Specifies the "A" Records array')
+param aRecords array = []
+
+@description('Optional. Specifies the "CNAME" Records array')
+param cnameRecords array = []
+
+@description('Optional. Specifies the "AAAA" Records array')
+param AAAARecords array = []
+
+resource dnszones 'Microsoft.Network/dnsZones@2018-05-01' = {
+  name: name
+  location: 'global'
+  tags: tags
+  properties: {
+    zoneType: zoneType
+  }
+}
+
+resource A 'Microsoft.Network/dnsZones/A@2018-05-01' = [for record in aRecords: {
+  name: record.name
+  parent: dnszones
+  properties: {
+    ARecords: (record.aliasRecordSet) ? null : record.aRecord
+    TTL: record.aliasRecordSet ? 0 : record.ttl
+    targetResource: record.aliasRecordSet ? {
+      id: record.targetResource
+    } : {}
+  }
+}]
+
+resource CName 'Microsoft.Network/dnsZones/CNAME@2018-05-01' = [for record in cnameRecords: {
+  name: record.name
+  parent: dnszones
+  properties: {
+    CNAMERecord: record.aliasRecordSet ? null : {
+      cname: record.cname
+    }
+    TTL: record.ttl
+    targetResource: record.aliasRecordSet ? {
+      id: record.targetResource
+    } : {}
+  }
+}]
+
+resource AAAA 'Microsoft.Network/dnsZones/AAAA@2018-05-01' = [for record in AAAARecords: {
+  name: record.name
+  parent: dnszones
+  properties: {
+    AAAARecords: (record.aliasRecordSet) ? null : record.aaaaRecord
+    TTL: record.ttl
+    targetResource: record.aliasRecordSet ? {
+      id: record.targetResource
+    } : {}
+  }
+}]
```

```suggestion
resource dnszones 'Microsoft.Network/dnsZones@2018-05-01' = {
  name: name
  location: 'global'
  tags: tags
  properties: {
    zoneType: zoneType
  }


  resource A 'A' = [for record in aRecords: {
    name: record.name
    properties: {
      ARecords: (record.aliasRecordSet) ? null : record.aRecord
      TTL: record.aliasRecordSet ? 0 : record.ttl
      targetResource: record.aliasRecordSet ? {
        id: record.targetResource
      } : {}
    }
  }]
  
  resource CName 'CNAME' = [for record in cnameRecords: {
    name: record.name
    properties: {
      CNAMERecord: record.aliasRecordSet ? null : {
        cname: record.cname
      }
      TTL: record.ttl
      targetResource: record.aliasRecordSet ? {
        id: record.targetResource
      } : {}
    }
  }]
  
  resource AAAA 'AAAA' = [for record in AAAARecords: {
    name: record.name
    properties: {
      AAAARecords: (record.aliasRecordSet) ? null : record.aaaaRecord
      TTL: record.ttl
      targetResource: record.aliasRecordSet ? {
        id: record.targetResource
      } : {}
    }
  }]
}
```

go with nested resource declaration when able.

---

any reasons for nested resource declaration? 
does current change I have to make it nested resource declaration?

---

the main thing I like about it is it reduced the instances of the API-Version to a single place. It also saves a few lines. I find it a little easier to read.
But not a big deal.

---

oh, reduced the instances of the API-Version that's a big deal, really like nested resource declaration.